### ISSUE TITLE
NodeIP:NodePort not accessible from outside cluster

### ISSUE THREAD

**Issue by nitendra-thakur** [cilium/cilium#26882]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

K8S Version: v1.24.2
Cilium Helm Chart Version: 1.12.6
Values override are:
```
k8sServiceHost: <k8s-api-fqdn>
k8sServicePort: 443
kubeProxyReplacement: strict

ipam:
  mode: "cluster-pool"
  operator:
    clusterPoolIPv4PodCIDR: "192.168.0.0/17"
    clusterPoolIPv4MaskSize: 24
```

K8S Nodes are on internal company network and are assigned a non-rfc-1918 IP from 100.0.0.0/16 (Not the actual range)
Cluster pool pod and service ranges (192.168.0.0/16) are not routable outside the cluster. Pods uses SNAT t talk outside. And outside clients must use nodeip and nodeport to access services.

I've deployed a nodeport service:
```
$ kubectl -n ingress-nginx get svc
NAME                             TYPE        CLUSTER-IP        EXTERNAL-IP   PORT(S)                      AGE
ingress-controller-controller    NodePort    192.168.178.105   <none>        80:31140/TCP,443:31130/TCP   3d4h
```

A curl test using NopeIP:NodePort (100.0.0.10:31140)
Works from pods.
Works from all the nodes of the cluster.
But DO NOT work from outside the cluster.

I intend to use an external HAProxy load balancer to access the nginx service, but cannot do that due to the above issue.

### Cilium Version

1.12.6

### Kernel Version

5.4.0-29

### Kubernetes Version

v1.24.2

### Relevant log output

```shell
external-haproxy-01:~$ curl 100.100.0.0.10:31140
curl: (7) Failed to connect to 100.0.0.10 port 31140: Connection refused


k8s-node-01:~$ curl 100.100.0.0.10:31140
<html>
<head><title>404 Not Found</title></head>
<body>
<center><h1>404 Not Found</h1></center>
<hr><center>nginx</center>
</body>
</html>


$ kubectl -n kube-system exec ds/cilium -- cilium status
Defaulted container "cilium-agent" out of: cilium-agent, mount-cgroup (init), apply-sysctl-overwrites (init), mount-bpf-fs (init), clean-cilium-state (init)
KVStore:                 Ok   Disabled
Kubernetes:              Ok   1.24 (v1.24.2) [linux/amd64]
Kubernetes APIs:         ["cilium/v2::CiliumClusterwideNetworkPolicy", "cilium/v2::CiliumEndpoint", "cilium/v2::CiliumNetworkPolicy", "cilium/v2::CiliumNode", "core/v1::Namespace", "core/v1::Node", "core/v1::Pods", "core/v1::Service", "discovery/v1::EndpointSlice", "networking.k8s.io/v1::NetworkPolicy"]
KubeProxyReplacement:    Strict   [rvni0 100.0.0.06]
Host firewall:           Disabled
CNI Chaining:            none
Cilium:                  Ok   1.12.6 (v1.12.6-9cc8d71)
NodeMonitor:             Listening for events on 16 CPUs with 64x4096 of shared memory
Cilium health daemon:    Ok
IPAM:                    IPv4: 2/254 allocated from 192.168.0.0/24,
ClusterMesh:             0/0 clusters ready, 0 global-services
BandwidthManager:        Disabled
Host Routing:            Legacy
Masquerading:            IPTables [IPv4: Enabled, IPv6: Disabled]
Controller Status:       19/19 healthy
Proxy Status:            OK, ip 192.168.0.128, 0 redirects active on ports 10000-20000
Global Identity Range:   min 256, max 65535
Hubble:                  Ok   Current/Max Flows: 4095/4095 (100.00%), Flows/s: 2.44   Metrics: Disabled
Encryption:              Disabled
Cluster health:          9/9 reachable   (2023-07-18T00:11:30Z)
```

### Anything else?

tcpdump on a k8s-node (100.0.0.10) when a `curl 100.0.0.10:31140` was done from outside the cluster; from a machine (origin-host-fqdn). I've replaced the actual hostnames and IPs n the following outputs.

```
k8s-node$ tcpdump -i any port 31140
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on any, link-type LINUX_SLL (Linux cooked v1), capture size 262144 bytes
23:56:01.561296 ethertype IPv4, IP <origin-host-fqdn>.57140 > <dest-k8s-node-fqdn>.31140: Flags [S], seq 1236660915, win 62720, options [mss 8960,sackOK,TS val 2845863714 ecr 0,nop,wscale 7], length 0
23:56:01.561296 IP <origin-host-fqdn>.57140 > <dest-k8s-node-fqdn>.31140: Flags [S], seq 1236660915, win 62720, options [mss 8960,sackOK,TS val 2845863714 ecr 0,nop,wscale 7], length 0
23:56:01.561408 IP <dest-k8s-node-fqdn>.31140 > <origin-host-fqdn>.57140: Flags [R.], seq 0, ack 1236660916, win 0, length 0
23:56:01.561415 ethertype IPv4, IP <dest-k8s-node-fqdn>.31140 > <origin-host-fqdn>.57140: Flags [R.], seq 0, ack 1, win 0, length 0
^C
4 packets captured
4 packets received by filter
0 packets dropped by kernel
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by aditighag** [comment]
Hi, Are you running Cilium with the kube-proxy replacement enabled or disabled? Also, please attach a sysdump.

**Comment by nitendra-thakur** [comment]
kube-proxy replacement is enabled. As you can see in the helm values override and cilium status that I had shared. `KubeProxyReplacement` is set to `strict`.
Attached here is the cilium sysdymp. But I had to remove the k8s yaml manifests as they contained internal workload details which may not be okay to share outside. I've also changed the k8s api endpoint fqdn and helm release name in existing cilium-*.yaml manifests. Hope that's okay; if not; let me know I can share any specific details that might help narrow down the issue.
[cilium-sysdump-20230718-155404.zip](https://github.com/cilium/cilium/files/12089631/cilium-sysdump-20230718-155404.zip)

**Comment by tamilmani1989** [comment]
@nitendra-thakur As you mentioned "K8S Nodes are on internal company network", is external client belong to same network?

**Comment by aditighag** [comment]
There is nothing in the sysdump for the IP you've mentioned in the original report. 

```
$ grep -r 100.0.0.10 * | wc -l
       0
```

**Comment by nitendra-thakur** [comment]
@aditighag `100.0.0.10` is not the actual IP. Like I said I had to change/mask those details in respect to company privacy. And yes, external client belongs to the same network.

I was debugging this issue further and came across a few observations:

All servers on this network (k8s nodes and the external clients) are using dual bgp peering with two ToR switches. There exists two interfaces on each node corresponding to these peerings. It seems that cilium does not auto detect these interfaces and so do not attach any ebpf programs to them.  
The same pair of interfaces (named eth101 and eth102) are also used for routing external traffic and host-to-host communication. Since there is no ebpf attachment to these interfaces (and of course no iptable rules), there is nothing that can respond to the request that are coming to the nodes with dest port as the nodeport. Thus it make sense why clients see "connection refused".

Issue [#23444](https://github.com/cilium/cilium/issues/23444) might be related to this.

I tried setting these interfaces under `devices` parameter that cilium-agent uses. But then it broke the whole routing on the nodes making them completely unreachable. After checking through bmc console, I noticed that all ip routes on the host were gone.

It might just be the right set parameters that I need to pass to cilium-agent. So far, I've tried ipv4NativeRoutingCIDR, nodePort.directRoutingDevice with no success.

**Comment by tamilmani1989** [comment]
would you able to reach from `node1` to `node2:nodeport`

**Comment by nitendra-thakur** [comment]
@tamilmani1989 
Yes, node to node it works. Only from outside the cluster, it does not. Here is a difference in tcpdump that was captured on node2, when the nodeport was accessed from node1 vs when it was access from an external client machine.
In this case, the nodeport assigned was 30425
```
$ kubectl --kubeconfig test-k8s -n nitest get svc
NAME    TYPE       CLUSTER-IP        EXTERNAL-IP   PORT(S)          AGE
nginx   NodePort   192.168.137.227   <none>        8080:30425/TCP   17h
```

<b>Scenario1: When nodeport was accessed from another node</b>
```
node1:~$ curl node2:30425
OK
```
tcpdump captured on node2
```
node2:~$ tcpdump -i any src node1-IP and dst node2-IP
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on any, link-type LINUX_SLL (Linux cooked v1), capture size 262144 bytes
18:29:12.092609 ethertype IPv4, IP node1-FQDN.46154 > node2-FQDN.8472: OTV, flags [I] (0x08), overlay 0, instance 6
IP 192.168.0.118.40050 > 192.168.1.123.http-alt: Flags [S], seq 4166161941, win 62370, options [mss 8910,sackOK,TS val 3962307732 ecr 0,nop,wscale 7], length 0
18:29:12.092609 IP node1-FQDN.46154 > node2-FQDN.8472: OTV, flags [I] (0x08), overlay 0, instance 6
IP 192.168.0.118.40050 > 192.168.1.123.http-alt: Flags [S], seq 4166161941, win 62370, options [mss 8910,sackOK,TS val 3962307732 ecr 0,nop,wscale 7], length 0
18:29:12.093118 ethertype IPv4, IP node1-FQDN.46154 > node1-FQDN.8472: OTV, flags [I] (0x08), overlay 0, instance 6
IP 192.168.0.118.40050 > 192.168.1.123.http-alt: Flags [.], ack 1057508688, win 488, options [nop,nop,TS val 3962307733 ecr 1213385637], length 0
18:29:12.093118 IP node1-FQDN.46154 > node2-FQDN.8472: OTV, flags [I] (0x08), overlay 0, instance 6
IP 192.168.0.118.40050 > 192.168.1.123.http-alt: Flags [.], ack 1, win 488, options [nop,nop,TS val 3962307733 ecr 1213385637], length 0
18:29:12.093204 ethertype IPv4, IP node1-FQDN.46154 > node2-FQDN.8472: OTV, flags [I] (0x08), overlay 0, instance 6
IP 192.168.0.118.40050 > 192.168.1.123.http-alt: Flags [P.], seq 0:85, ack 1, win 488, options [nop,nop,TS val 3962307733 ecr 1213385637], length 85: HTTP: GET / HTTP/1.1
18:29:12.093204 IP node1-FQDN.46154 > node1-FQDN.8472: OTV, flags [I] (0x08), overlay 0, instance 6
IP 192.168.0.118.40050 > 192.168.1.123.http-alt: Flags [P.], seq 0:85, ack 1, win 488, options [nop,nop,TS val 3962307733 ecr 1213385637], length 85: HTTP: GET / HTTP/1.1
18:29:12.093737 ethertype IPv4, IP node1-FQDN.46154 > node1-FQDN.8472: OTV, flags [I] (0x08), overlay 0, instance 6
IP 192.168.0.118.40050 > 192.168.1.123.http-alt: Flags [.], ack 233, win 487, options [nop,nop,TS val 3962307733 ecr 1213385637], length 0
18:29:12.093737 IP node1-FQDN.46154 > node1-FQDN.8472: OTV, flags [I] (0x08), overlay 0, instance 6
IP 192.168.0.118.40050 > 192.168.1.123.http-alt: Flags [.], ack 233, win 487, options [nop,nop,TS val 3962307733 ecr 1213385637], length 0
^C
8 packets captured
12 packets received by filter
4 packets dropped by kernel
```


<b>Scenario2: When nodeport was accessed from external client machine</b>
```
client:~$ curl node2:30425
[OK](curl: (7) Failed to connect to node2-FQDN port 30425 after 2 ms: Connection refused)
```
tcpdump captured on node2
```
tcpdump -i any src client-IP and dst node2-IP
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on any, link-type LINUX_SLL (Linux cooked v1), capture size 262144 bytes
18:37:08.563489 ethertype IPv4, IP client-FQDN.43680 > node2-FQDN.30425: Flags [S], seq 359595177, win 62720, options [mss 8960,sackOK,TS val 1179198645 ecr 0,nop,wscale 7], length 0
18:37:08.563489 IP client-FQDN.43680 > node2-FQDN.30425: Flags [S], seq 359595177, win 62720, options [mss 8960,sackOK,TS val 1179198645 ecr 0,nop,wscale 7], length 0
^C
2 packets captured
2 packets received by filter
0 packets dropped by kernel
```

One thing I noted was, there is no trace of nodeport 30425 in tcpdump of scenario1. If I put a filter of `port 30425` on tcpdump; I get nothing. May be cilium is changing the dest port on the source (node1) itself. It seems the overlay network created by cilium is what makes scenario work.

`192.168.0.0/17` is the clusterPoolIPv4PodCIDR used on the cluster. This range is not routable on the host network. And all traffic leaving the cluster must be SNATed using node ip addresses which falls under a different ip range (for example 100.0.0.0/8). 

This is the iptables rules created by cilium on node2:
```
node2:~# iptables-save | grep CILIUM
:CILIUM_OUTPUT_raw - [0:0]
:CILIUM_PRE_raw - [0:0]
-A PREROUTING -m comment --comment "cilium-feeder: CILIUM_PRE_raw" -j CILIUM_PRE_raw
-A OUTPUT -m comment --comment "cilium-feeder: CILIUM_OUTPUT_raw" -j CILIUM_OUTPUT_raw
-A CILIUM_OUTPUT_raw -o lxc+ -m mark --mark 0xa00/0xfffffeff -m comment --comment "cilium: NOTRACK for proxy return traffic" -j CT --notrack
-A CILIUM_OUTPUT_raw -o cilium_host -m mark --mark 0xa00/0xfffffeff -m comment --comment "cilium: NOTRACK for proxy return traffic" -j CT --notrack
-A CILIUM_OUTPUT_raw -o lxc+ -m mark --mark 0x800/0xe00 -m comment --comment "cilium: NOTRACK for L7 proxy upstream traffic" -j CT --notrack
-A CILIUM_OUTPUT_raw -o cilium_host -m mark --mark 0x800/0xe00 -m comment --comment "cilium: NOTRACK for L7 proxy upstream traffic" -j CT --notrack
-A CILIUM_PRE_raw -m mark --mark 0x200/0xf00 -m comment --comment "cilium: NOTRACK for proxy traffic" -j CT --notrack
:CILIUM_OUTPUT_nat - [0:0]
:CILIUM_POST_nat - [0:0]
:CILIUM_PRE_nat - [0:0]
-A PREROUTING -m comment --comment "cilium-feeder: CILIUM_PRE_nat" -j CILIUM_PRE_nat
-A OUTPUT -m comment --comment "cilium-feeder: CILIUM_OUTPUT_nat" -j CILIUM_OUTPUT_nat
-A POSTROUTING -m comment --comment "cilium-feeder: CILIUM_POST_nat" -j CILIUM_POST_nat
-A CILIUM_POST_nat -s 192.168.1.0/24 ! -d 192.168.1.0/24 ! -o cilium_+ -m comment --comment "cilium masquerade non-cluster" -j MASQUERADE
-A CILIUM_POST_nat -m mark --mark 0xa00/0xe00 -m comment --comment "exclude proxy return traffic from masquerade" -j ACCEPT
-A CILIUM_POST_nat ! -s 192.168.1.0/24 ! -d 192.168.1.0/24 -o cilium_host -m comment --comment "cilium host->cluster masquerade" -j SNAT --to-source 192.168.1.224
-A CILIUM_POST_nat -s 127.0.0.1/32 -o cilium_host -m comment --comment "cilium host->cluster from 127.0.0.1 masquerade" -j SNAT --to-source 192.168.1.224
-A CILIUM_POST_nat -o cilium_host -m mark --mark 0xf00/0xf00 -m conntrack --ctstate DNAT -m comment --comment "hairpin traffic that originated from a local pod" -j SNAT --to-source 192.168.1.224
:CILIUM_FORWARD - [0:0]
:CILIUM_INPUT - [0:0]
:CILIUM_OUTPUT - [0:0]
-A INPUT -m comment --comment "cilium-feeder: CILIUM_INPUT" -j CILIUM_INPUT
-A FORWARD -m comment --comment "cilium-feeder: CILIUM_FORWARD" -j CILIUM_FORWARD
-A OUTPUT -m comment --comment "cilium-feeder: CILIUM_OUTPUT" -j CILIUM_OUTPUT
-A CILIUM_FORWARD -o cilium_host -m comment --comment "cilium: any->cluster on cilium_host forward accept" -j ACCEPT
-A CILIUM_FORWARD -i cilium_host -m comment --comment "cilium: cluster->any on cilium_host forward accept (nodeport)" -j ACCEPT
-A CILIUM_FORWARD -i lxc+ -m comment --comment "cilium: cluster->any on lxc+ forward accept" -j ACCEPT
-A CILIUM_FORWARD -i cilium_net -m comment --comment "cilium: cluster->any on cilium_net forward accept (nodeport)" -j ACCEPT
-A CILIUM_INPUT -m mark --mark 0x200/0xf00 -m comment --comment "cilium: ACCEPT for proxy traffic" -j ACCEPT
-A CILIUM_OUTPUT -m mark --mark 0xa00/0xfffffeff -m comment --comment "cilium: ACCEPT for proxy return traffic" -j ACCEPT
-A CILIUM_OUTPUT -m mark --mark 0x800/0xe00 -m comment --comment "cilium: ACCEPT for l7 proxy upstream traffic" -j ACCEPT
-A CILIUM_OUTPUT -m mark ! --mark 0xe00/0xf00 -m mark ! --mark 0xd00/0xf00 -m mark ! --mark 0xa00/0xe00 -m mark ! --mark 0x800/0xe00 -m mark ! --mark 0xf00/0xf00 -m comment --comment "cilium: host->any mark as from host" -j MARK --set-xmark 0xc00/0xf00
:CILIUM_POST_mangle - [0:0]
:CILIUM_PRE_mangle - [0:0]
-A PREROUTING -m comment --comment "cilium-feeder: CILIUM_PRE_mangle" -j CILIUM_PRE_mangle
-A POSTROUTING -m comment --comment "cilium-feeder: CILIUM_POST_mangle" -j CILIUM_POST_mangle
-A CILIUM_PRE_mangle -m socket --transparent -m comment --comment "cilium: any->pod redirect proxied traffic to host proxy" -j MARK --set-xmark 0x200/0xffffffff
-A CILIUM_PRE_mangle -p tcp -m mark --mark 0xd7810200 -m comment --comment "cilium: TPROXY to host cilium-dns-egress proxy" -j TPROXY --on-port 33239 --on-ip 0.0.0.0 --tproxy-mark 0x200/0xffffffff
-A CILIUM_PRE_mangle -p udp -m mark --mark 0xd7810200 -m comment --comment "cilium: TPROXY to host cilium-dns-egress proxy" -j TPROXY --on-port 33239 --on-ip 0.0.0.0 --tproxy-mark 0x200/0xffffffff
```

**Comment by tamilmani1989** [comment]
ya you are right if node is within cluster then translation happens in src side.. Let's check if any ebpf program attached on your VM primary nic.. can you share output of this running on your VM primary interface?
```
tc filter show dev <vm_primary_nic> ingress
tc filter show dev <vm_primary_nic> egress
```
should result in something like this:
filter protocol all pref 1 bpf chain 0 handle 0x1 **cil_from_netdev-eth0** direct-action not_in_hw id 1099 tag 95e69f5eb12703a8 jited
filter protocol all pref 1 bpf chain 0 handle 0x1 **cil_to_netdev-eth0** direct-action not_in_hw id 1113 tag b3fa58ac8199d695 jited

**Comment by nitendra-thakur** [comment]
Thank you @tamilmani1989  and @aditighag  for your responses. I was able to figure out the issue and come up with a workaround for us. However a cleaner solution would require a fix in auto device detection logic. I've raised a separate issue to track that on. https://github.com/cilium/cilium/issues/27121

Let me summarize here on this one, and we can close this issue.
* Auto device detection did not work since we are using ecmp on all k8s nodes
* We are also using sub interfaces which are attached as vlan to the physical interface
* When I used `--devices` options, I was wrongly specifying the physical interfaces. The bpf attached to the physical interface would then override all the dynamic routes received on vlans via bgp. The simple solution was to specify sub-interfaces; and not the physical interfaces.
